### ISSUE TITLE
Tensor.gradient should be a function (method)

### ISSUE THREAD

**Issue by geohot** [geohot/tinygrad#7183]
`grad_weights = loss.gradient(*weights)`

Should confirm this API works for second derivatives too, but I see no reason why it shouldn't.

Blocked on big graph where the grad will actually be computed, need LazyBuffer to be UOp. Also, instead of view, we can bring back MovementOps so we can compute derivative. Derivative computation will be done in a pattern matcher.

`requires_grad` will be removed as well as function.py

Example will become:
```python
from tinygrad import Tensor

x = Tensor.eye(3)
y = Tensor([[2.0,0,-2.0]])
z = y.matmul(x).sum()
x_grad, y_grad = z.gradient(x, y)

print(x_grad.numpy())  # dz/dx
print(y_grad.numpy())  # dz/dy
```

I think `gradient` feels more natural than `grad`, and as a bonus we don't have to break the old API, `x.grad` becomes an accessor which calls `backward_tensor.gradient(x)`

**Comment by geohot** [comment]
It's interesting how this plays into the UOp freeing problem. If there's no require_grad, you don't know how much of the graph to keep. Is it possible to keep it all, and do something like dynamic bound assign?

**Comment by radenmuaz** [comment]
I made something similar, basically extending jax autodidax tutorial to something more fleshed out, but line count got pretty high.
https://github.com/radenmuaz/slope-ad
